### PR TITLE
feat(liquidator): update liquidator tests with non 18 decimal cases

### LINKED ISSUE #1902: Upgrade pricefeedmock for non 18 decimal and update liquidator bot
**Why**

Want to continue the effort to support non 18 decimal collateral currencies, specifically targeting liquidator bot tests ( and maybe dispute bot). Liquidator bot tests currently use uniswap mock, but should use the pricefeedmock with non 18 decimal support. 

Breaking this out as a seperate PR so that previous work can be reviewed and merged. 

**How**

- Update pricefeedmock to behave correctly for non 18 decimal currency, similar to[ recent cryptowatch updates](https://github.com/UMAprotocol/protocol/pull/1885).
- Update liquidator bot tests from uniswap to pricefeedmock
- Update dispute bot tests in same way

**Resolution**

See that liquidator/dispute bot tests pass with non 18 dec collateral currency. 

**Difficulty Score [1-10]**

7

### PR DESCRIPTION
**Pull request by chrismaree**
**Motivation**

#1902 

This PR replaces [1913](https://github.com/UMAprotocol/protocol/pull/1913)

**Summary**

Updated pricefeed mock to do conversions based on `decimals` provided in constructor. This wont change previous behavior as they will default to 18. Old tests are left in to verify nothing breaks. The tests are completely duplicated to tests for 8 decimal cases, and values are converted where appropriate.  A new issue/ PR will be opened for disputer bot changes.

For some reason diffs on the test file are completely messed up. I had to add a describe block for 18 decimal as well as 8 decimal test cases, and that probably screwed it up.

**Issue(s)**

Fixes #1902

### CODE REVIEW

File: packages/common/src/FormattingUtils.js [hunk 1]
```diff
@@ -117,6 +119,30 @@ function addSign(number) {
...
 
+// Take an amount based on fromDecimals and convert it to an amount based on toDecimals
+// For example amount = 100 usdt, its decimals are 6. You want to convert it to 18.
+// convertDecimals(6,18)(100000000)  => 1000000000000000000000.
+// Returns a BigNumber you will need to call toString on
+// fromDecimals: number - decimal value of amount
+// toDecimals: number - decimal value to convert to
+// bn: toBN function - optionally provide your own BN function.
+// return => (amount:string)=>BN
+const ConvertDecimals = (fromDecimals, toDecimals, bn = toBN) => {
```
Comment (nicholaspai) [inline packages/common/src/FormattingUtils.js:130]: OOC why is this different than parseFixed and formatFixed

File: packages/financial-templates-lib/src/clients/ExpiringMultiPartyClient.js [hunk 2]
```diff
@@ -1,8 +1,10 @@
 // A thick client for getting information about an ExpiringMultiParty. Used to get sponsor information, outstanding
 // positions, undisputed Liquidations, expired liquidations, disputed liquidations.
 
-const { LiquidationStatesEnum } = require("@umaprotocol/common");
+const { ConvertDecimals, parseFixed, LiquidationStatesEnum } = require("@umaprotocol/common");
 const Promise = require("bluebird");
+// const { parseFixed } = require("@ethersproject/bignumber");
```
Comment (nicholaspai) [inline packages/financial-templates-lib/src/clients/ExpiringMultiPartyClient.js:6]: Forgot to remove these comments

File: packages/financial-templates-lib/test/clients/ExpiringMultiPartyClient.js [hunk 3]
```diff
@@ -14,494 +15,519 @@ const TokenFactory = artifacts.require("TokenFactory");
...
-
-    // Force a price such that the dispute fails, and then
-    // withdraw from the unsuccessfully disputed liquidation and check that the liquidation is deleted.
-    const disputePrice = toWei("1.6");
-    await mockOracle.pushPrice(web3.utils.utf8ToHex("UMATEST"), liquidationTime, disputePrice);
-    await emp.withdrawLiquidation("0", sponsor1, { from: liquidator });
-    await client.update();
-    assert.deepStrictEqual([], client.getDisputedLiquidations().sort());
+contract("ExpiringMultiPartyClient.js", function(accounts) {
+  configs.forEach(({ collateralDecimals, tokenName }) => {
```
Comment (nicholaspai) [inline packages/financial-templates-lib/test/clients/ExpiringMultiPartyClient.js:26]: As discussed offline, I'd change this from `forEach` to a `for(let x of X)` or `for(let x; x < len; x++)`. For each sometimes gives unexpected results for asynchronous code: https://stackoverflow.com/questions/37576685/using-async-await-with-a-foreach-loop#37576787

File: packages/common/src/FormattingUtils.js (hunk 1 above)
Comment (daywiss) [reply packages/common/src/FormattingUtils.js:130]: this was needed for a call. converting from synthetic decimals to collateral decimals in the emp.

File: packages/liquidator/test/Liquidator.js [hunk 4]
```diff
@@ -31,818 +36,145 @@ const TokenFactory = artifacts.require("TokenFactory");
...
-    await collateralToken.mint(liquidatorBot, toWei("100000"), { from: contractCreator });
+const Convert = decimals => number => parseFixed(number.toString(), decimals).toString();
 
-    // Create identifier whitelist and register the price tracking ticker with it.
-    identifierWhitelist = await IdentifierWhitelist.deployed();
-    await identifierWhitelist.addSupportedIdentifier(web3.utils.utf8ToHex("UMATEST"));
-  });
+contract("Liquidator.js", function(accounts) {
+  configs.forEach(({ collateralDecimals, tokenName }) => {
+    describe(`${collateralDecimals} decimals`, function() {
```
Comment (nicholaspai) [inline packages/liquidator/test/Liquidator.js:48]: Same comment as above

**Review by nicholaspai — APPROVED** [review]
Looks great, I think you can change forEach to for loop to be async/sync safe.

File: packages/financial-templates-lib/test/clients/ExpiringMultiPartyClient.js (hunk 3 above)
Comment (mrice32) [reply packages/financial-templates-lib/test/clients/ExpiringMultiPartyClient.js:26]: Is there any async code being called in this for each, though? Or are you just saying it's safer to preempt any issues?

File: packages/liquidator/test/Liquidator.js [hunk 5]
```diff
@@ -855,168 +187,892 @@ contract("Liquidator.js", function(accounts) {
...
+        await liquidator.update();
+        await liquidator.liquidatePositions();
+        assert.equal(spy.callCount, 2); // 1 new info level event due to the successful liquidation.
+
+        // The liquidation should have gone through.
+        assert.equal((await emp.getLiquidations(sponsor1)).length, 1);
+        assert.equal(spy.callCount, 2); // 1 new log level event due to the successful execution.
+      });
+
+      it("amount-to-liquidate > min-sponsor-tokens, swaps reserveToken (ETH) for tokenCurrency on OneSplit", async function() {
```
Comment (mrice32) [inline packages/liquidator/test/Liquidator.js:540]: nit: why did this test move?

File: packages/common/src/FormattingUtils.js [hunk 6]
```diff
@@ -3,6 +3,8 @@ const networkUtils = require("./PublicNetworks");
 const BigNumber = require("bignumber.js");
 const moment = require("moment");
 const { formatFixed, parseFixed } = require("@ethersproject/bignumber");
+const web3 = require("web3");
+const { toBN } = web3.utils;
```
Comment (mrice32) [inline packages/common/src/FormattingUtils.js:7]: nit: it's idiomatic (when using web3) to name the _library_ `Web3` and the instance (result of `new Web3(provider)`) `web3`.

```suggestion
const Web3 = require("web3");
const { toBN } = Web3.utils;
```

File: packages/financial-templates-lib/test/clients/ExpiringMultiPartyClient.js (hunk 3 above)
Comment (chrismaree) [reply packages/financial-templates-lib/test/clients/ExpiringMultiPartyClient.js:26]: I have updated to your spect @nicholaspai

File: packages/common/src/FormattingUtils.js (hunk 6 above)
Comment (chrismaree) [reply packages/common/src/FormattingUtils.js:7]: I chose to remove this in full and inject the `web3` object like we do with the other formatters in this file.

**Review by mrice32 — APPROVED** [review]
LGTM!

**Pull request merged**